src/Arrays/MPIStateArrays.jl

```diff
@@ -420,49 +382,127 @@ function finish_ghost_send!(Q::MPIStateArray)
-            sendbuf,
-            buf,
-            vmapsend,
-            length(vmapsend);
-            ndrange = length(vmapsend),
-            dependencies = (event,),
+function __testall!(requests)
+    done = false
+    while !done
+        done, _ = MPI.Testall!(requests)
+        yield()
+    end
+end
+
+function __Irecv!(Q)
+    nnabr = length(Q.nabrtorank)
+    transfer = get_transfer(Q.recv_buffer)
+
+    for n in 1:nnabr
+        # If this fails we haven't waited on previous recv!
+        @assert Q.recvreq[n].buffer == nothing
+
+        Q.recvreq[n] = MPI.Irecv!(
+            (@view transfer[:, Q.nabrtovmaprecv[n]]),
+            Q.nabrtorank[n],
+            666,
+            Q.mpicomm,
         )
-        wait(device(buf), event)
     end
 end
 
-function transferrecvbuf!(buf, recvbuf, vmaprecv)
-    if length(vmaprecv) > 0
-        Np = size(buf, 1)
-        nvar = size(buf, 2)
+function __Isend!(Q)
+    nnabr = length(Q.nabrtorank)
+    transfer = get_transfer(Q.send_buffer)
 
-        event = Event(device(buf))
-        event = knl_transferrecvbuf!(device(buf), 256)(
-            Val(Np),
-            Val(nvar),
-            buf,
-            recvbuf,
-            vmaprecv,
-            length(vmaprecv);
-            ndrange = length(vmaprecv),
-            dependencies = (event,),
+    for n in 1:nnabr
+        Q.sendreq[n] = MPI.Isend(
+            (@view transfer[:, Q.nabrtovmapsend[n]]),
+            Q.nabrtorank[n],
+            666,
+            Q.mpicomm,
         )
-        wait(device(buf), event)
     end
 end
 
+# {{{ MPI Buffer handling
+function fillsendbuf!(sendbuf, buf, vmapsend; dependencies = nothing)
+    if vmapsend == 0
+        return MultiEvent(dependencies)
+    end
+
+    Np = size(buf, 1)
+    nvar = size(buf, 2)
+
+    event = knl_fillsendbuf!(device(buf), 256)(
+        Val(Np),
+        Val(nvar),
+        sendbuf,
+        buf,
+        vmapsend,
+        length(vmapsend);
+        ndrange = length(vmapsend),
+        dependencies = dependencies,
+    )
+
+    return event
+end
+
+function transferrecvbuf!(buf, recvbuf, vmaprecv; dependencies = nothing)
+    if vmaprecv == 0
+        return MultiEvent(dependencies)
+    end
+
+    Np = size(buf, 1)
+    nvar = size(buf, 2)
+
+    event = knl_transferrecvbuf!(device(buf), 256)(
+        Val(Np),
+        Val(nvar),
+        buf,
+        recvbuf,
+        vmaprecv,
+        length(vmaprecv);
+        ndrange = length(vmaprecv),
+        dependencies = dependencies,
+    )
+
+    return event
+end
+
 # }}}
 
+"""
+    ghost_exchange!(Q::MPIStateArray; dependencies=nothing)
+
+Asynchronous call to perform the exchange of ghost degrees-of-freedom.  The
+returned `KernelAbstractions.Event` can be waited on to know when the ghost
+exchange is finished.
+"""
+function ghost_exchange!(Q::MPIStateArray; dependencies = nothing)
+    event = Event(__Irecv!, Q; dependencies = dependencies)
+    event = Event(__testall!, Q.recvreq; dependencies = event)
```

Comment:
This means we need to make sure that MPI is using the thread safe interface, right now this interface is free to execute on any thread. https://github.com/JuliaGPU/KernelAbstractions.jl/blob/8a95e72e286d77be15de75971c4f52eaa4b8a599/src/backends/cpu.jl#L9-L15

@kpamnany do you have a feeling for what the performance impacts are these days for `MPI_Init_thread` with `MPI_THREAD_MULTIPLE`? We will also need to add support for that in MPI.jl

Reply:
Do we want to have an option to `Event` to run in the main thread?

Reply:
I've added support for `MPI.Init_thread` in https://github.com/JuliaParallel/MPI.jl/pull/363